streams/shared/src/main/scala/zio/stream/ZPipeline.scala

```diff
@@ -228,6 +228,104 @@ object ZPipeline extends ZPipelinePlatformSpecificConstructors {
+
+          lazy val producer: ZChannel[Env, Err, Chunk[In], Any, Err, Nothing, Any] =
+            ZChannel.readWithCause(
+              (in: Chunk[In]) =>
+                in.lastOption.fold(producer) { last =>
+                  ZChannel.fromZIO(handoff.offer(Emit(Chunk.single(last)))) *> producer
+                },
+              (cause: Cause[Err]) => ZChannel.fromZIO(handoff.offer(Halt(cause))),
+              (_: Any) => ZChannel.fromZIO(handoff.offer(End(ZStream.SinkEndReason.UpstreamEnd)))
+            )
+
+          def consumer(state: DebounceState[Err, In]): ZChannel[Env, Any, Any, Any, Err, Chunk[In], Any] =
+            ZChannel.unwrap(
+              state match {
+                case NotStarted =>
+                  handoff.take.map {
+                    case Emit(last) =>
+                      ZChannel.unwrap(enqueue(last))
+                    case HandoffSignal.Halt(error) =>
+                      ZChannel.failCause(error)
+                    case HandoffSignal.End(_) =>
+                      ZChannel.unit
+                  }
+                case Current(fiber) =>
+                  fiber.join.map {
+                    case HandoffSignal.Emit(last) => ZChannel.unwrap(enqueue(last))
+                    case HandoffSignal.Halt(error) => ZChannel.failCause(error)
+                    case HandoffSignal.End(_) => ZChannel.unit
+                  }
+                case Previous(fiber) =>
+                  fiber.join
+                    .raceWith[Env, Err, Err, HandoffSignal[Err, In], ZChannel[
+                      Env,
+                      Any,
+                      Any,
+                      Any,
+                      Err,
+                      Chunk[In],
+                      Any
+                    ]](
+                      handoff.take
+                    )(
+                      {
+                        case (Exit.Success(a), current) =>
+                          ZIO.succeedNow(ZChannel.write(a) *> consumer(Current(current)))
+                        case (Exit.Failure(cause), current) =>
+                          current.interrupt as ZChannel.failCause(cause)
+                      },
+                      {
+                        case (Exit.Success(Emit(last)), previous) =>
+                          previous.interrupt *> enqueue(last)
+                        case (Exit.Success(Halt(cause)), previous) =>
+                          previous.interrupt as ZChannel.failCause(cause)
+                        case (Exit.Success(End(_)), previous) =>
+                          previous.join.map(ZChannel.write(_) *> ZChannel.unit)
+                        case (Exit.Failure(cause), previous) =>
+                          previous.interrupt as ZChannel.failCause(cause)
+                      }
+                    )
+              }
+            )
+
+          ZChannel.scoped[Env](producer.runScoped.forkScoped) *>
+            new ZPipeline(consumer(NotStarted))
+        }
+      }
+    )
+  }
```

Comment:
I haven't checked it locally but I would suggest running everything in `ZChannel` and only then wrapping it in a `ZPipeline`

i.e.

```scala
def debounce[Env, Err, In](d: => Duration)(implicit trace: Trace): ZPipeline[Env, Err, In, In] = {
    import ZStream.DebounceState
    import ZStream.DebounceState._
    import ZStream.HandoffSignal
    import ZStream.HandoffSignal._

new ZPipeline(ZChannel.unwrap {
   // ...  

  ZChannel.unwrapScoped(producer.runScoped.forkScoped as consumer(NotStarted))
})

```